overlord/servicestate/quota_handlers.go

```diff
@@ -781,31 +802,41 @@ func ensureSnapServicesStateForGroup(st *state.State, grp *quota.Group, opts *en
 	return restartSnapServices(st, appsToRestartBySnap)
 }
 
-func ensureGroupIsNotMixed(group string, allGrps map[string]*quota.Group) error {
+// ensureGroupIsNotMixed is a helper which returns an error if the snaps and/or
+// services we are trying to add would result in a mixed group. A group is considered
+// mixed if:
+// 1. The group would end up containing both snaps and services
+// 2. The group would end up containing both services and groups
+func ensureGroupIsNotMixed(snapsToAdd, servicesToAdd []string, group string, allGrps map[string]*quota.Group) error {
+	if len(snapsToAdd) > 0 && len(servicesToAdd) > 0 {
+		return fmt.Errorf("cannot mix services and snaps in the same quota group")
+	}
+
 	grp, ok := allGrps[group]
-	if ok && len(grp.SubGroups) != 0 && len(grp.Snaps) != 0 {
-		return fmt.Errorf("quota group %q has mixed snaps and sub-groups, which is no longer supported; removal and re-creation is necessary to modify it", group)
+	if ok && ((len(grp.Services) > 0 && len(snapsToAdd) > 0) || (len(grp.Snaps) > 0 && len(servicesToAdd) > 0)) {
+		return fmt.Errorf("cannot mix services and snaps in the same quota group")
```

Comment:
this error case and other new errors later in the code are not reached by tests

Reply:
Added tests to cover this